floss/language/utils.py

```diff
@@ -0,0 +1,623 @@
+    to find all the 32-bit absolutely addressed LEA instructions,
+    extracting the target virtual address.
+    """
+    absolute_insn_re = re.compile(
+        rb"""
+        (
+              \x8D \x05  # 8d 05 aa aa 00 00       lea    eax,ds:0xaaaa
+            | \x8D \x1D  # 8d 1d aa aa 00 00       lea    ebx,ds:0xaaaa
+            | \x8D \x0D  # 8d 0d aa aa 00 00       lea    ecx,ds:0xaaaa
+            | \x8D \x15  # 8d 15 aa aa 00 00       lea    edx,ds:0xaaaa
+            | \x8D \x35  # 8d 35 aa aa 00 00       lea    esi,ds:0xaaaa
+            | \x8D \x3D  # 8d 3d aa aa 00 00       lea    edi,ds:0xaaaa
+        )
+        (?P<address>....)
+        """,
+        re.DOTALL + re.VERBOSE,
+    )
+
+    for match in absolute_insn_re.finditer(buf):
+        address_bytes = match.group("address")
+        address = struct.unpack("<I", address_bytes)[0]
+
+        yield address
+
+
+def find_lea_xrefs(pe: pefile.PE) -> Iterable[VA]:
+    """
+    scan the executable sections of the given PE file
+    for LEA instructions that reference valid memory addresses,
+    yielding the virtual addresses.
+    """
+    low, high = get_image_range(pe)
+
+    for section in pe.sections:
+        if not section.IMAGE_SCN_MEM_EXECUTE:
+            continue
+
+        code = section.get_data()
+
+        if pe.FILE_HEADER.Machine == pefile.MACHINE_TYPE["IMAGE_FILE_MACHINE_AMD64"]:
+            xrefs = find_amd64_lea_xrefs(code, section.VirtualAddress + pe.OPTIONAL_HEADER.ImageBase)
+        elif pe.FILE_HEADER.Machine == pefile.MACHINE_TYPE["IMAGE_FILE_MACHINE_I386"]:
+            xrefs = find_i386_lea_xrefs(code)
+        else:
+            raise ValueError("unhandled architecture")
+
+        for xref in xrefs:
+            if low <= xref < high:
+                yield xref
+
+
+def find_i386_push_xrefs(buf: bytes) -> Iterable[VA]:
+    """
+    scan the given data found at the given base address
+    to find all the 32-bit PUSH instructions,
+    extracting the target virtual address.
+    """
+    push_insn_re = re.compile(
+        rb"""
+        (
+              \x68       # 68 aa aa 00 00       push   0xaaaa
+        )
+        (?P<address>....)
+        """,
+        re.DOTALL + re.VERBOSE,
+    )
+
+    for match in push_insn_re.finditer(buf):
+        address_bytes = match.group("address")
+        address = struct.unpack("<I", address_bytes)[0]
+
+        yield address
+
+
+def find_push_xrefs(pe: pefile.PE) -> Iterable[VA]:
+    """
+    scan the executable sections of the given PE file
+    for PUSH instructions that reference valid memory addresses,
+    yielding the virtual addresses.
+    """
+    low, high = get_image_range(pe)
+
+    for section in pe.sections:
+        if not section.IMAGE_SCN_MEM_EXECUTE:
+            continue
+
+        code = section.get_data()
+
+        if pe.FILE_HEADER.Machine == pefile.MACHINE_TYPE["IMAGE_FILE_MACHINE_AMD64"]:
+            xrefs: Iterable[VA] = []  # no push instructions on amd64
```

Comment:
can we be sure string references would never be pushed? probably, but can you add some context/references around this? the comment is a little misleading currently